modules/indexing/src/main/java/org/apache/ignite/internal/processors/query/stat/IgniteGlobalStatisticsManager.java

```diff
@@ -0,0 +1,1026 @@
+            }
+        }
+        else
+            ioMgr.sendToGridTopic(nodeId, GridTopic.TOPIC_STATISTICS, msg, GridIoPolicy.MANAGEMENT_POOL);
+    }
+
+    /**
+     * Send statistics response or process it locally.
+     *
+     * @param nodeId Target node id. If equals to local node - corresponding method will be called directly.
+     * @param msg Statistics response to send.
+     * @throws IgniteCheckedException In case of error.
+     */
+    private void send(UUID nodeId, StatisticsResponse msg) throws IgniteCheckedException {
+        if (discoMgr.localNode().id().equals(nodeId)) {
+            switch (msg.data().type()) {
+                case LOCAL:
+                    processLocalResponse(nodeId, msg);
+
+                    break;
+
+                case GLOBAL:
+                    processGlobalResponse(nodeId, msg);
+
+                    break;
+
+                default:
+                    log.warning("Unexpected type " + msg.data().type() + " in statistics response message " + msg);
+            }
+        }
+        else
+            ioMgr.sendToGridTopic(nodeId, GridTopic.TOPIC_STATISTICS, msg, GridIoPolicy.MANAGEMENT_POOL);
+    }
+
+    /** Cache entry. */
+    private static class CacheEntry<T> {
+        /** Cached object. */
+        private final T obj;
+
+        /**
+         * Constructor.
+         *
+         * @param obj Cached object.
+         */
+        public CacheEntry(T obj) {
+            this.obj = obj;
+        }
+
+        /**
+         * @return Cached object.
+         */
+        public T object() {
+            return obj;
+        }
+    }
+
+    /** Context of global statistics gathering. */
+    private static class StatisticsGatheringContext {
+        /** Number of remaining requests. */
+        private int remainingResponses;
+
+        /** Requests id. */
+        private final UUID reqId;
+
+        /** Local object statistics from responses. */
+        private final Collection<ObjectStatisticsImpl> responses = new ArrayList<>();
+
+        /** Configuration, used to collect statistics. */
+        private final StatisticsObjectConfiguration cfg;
+
+        /**
+         * Constructor.
+         *
+         * @param responseCont Expectiong response count.
+         * @param reqId Requests id.
+         * @param cfg Configuration, used to collect statistics.
+         */
+        public StatisticsGatheringContext(int responseCont, UUID reqId, StatisticsObjectConfiguration cfg) {
+            remainingResponses = responseCont;
+            this.reqId = reqId;
+            this.cfg = cfg;
+        }
+
+        /**
+         * Register response.
+         *
+         * @param data Object statistics from response.
+         * @return {@code true} if all respones collected, {@code false} otherwise.
+         */
+        public synchronized boolean registerResponse(ObjectStatisticsImpl data) {
```

Comment:
why do you need this method sync?

Reply:
Just to use simple array list collections inside.